src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c

```diff
@@ -630,13 +639,58 @@ static void rr_connectivity_changed_locked(grpc_exec_ctx *exec_ctx, void *arg,
         GRPC_CLOSURE_SCHED(exec_ctx, pp->on_complete, GRPC_ERROR_NONE);
         gpr_free(pp);
       }
+      rr_subchannel_list_unref(exec_ctx, sd->subchannel_list,
+                               "sd_shutdown+started_picking");
+      // unref the "rr_connectivity_update" weak ref from start_picking.
+      GRPC_LB_POLICY_WEAK_UNREF(exec_ctx, &p->base,
+                                "rr_connectivity_sd_shutdown");
+    } else {
+      // Policy isn't shutting down, we aren't in connectivity shutdown (from
```

Comment:
I think I wasn't clear in expressing my concern here.  Let me try a different way.

The code in `update_state_counters_locked()` updates the counters based on `prev_connectivity_state` and `curr_connectivity_state`, and it assumes that it will be called whenever the state changes.  However, in this new code, when we try to reconnect, we reset `prev_connectivity_state` and `curr_connectivity_state` without calling `update_state_counters_locked()`, which means that we're not updating the counters appropriately.

In particular, the next time `update_state_counters_locked()` is called, `prev_connectivity_state` will no longer be set to `SHUTDOWN`, so `num_shutdown` will never be decremented, which means that it will grow until we have seen the same number of shutdowns as the number of subchannels.  For example, let's say that we have 3 subchannels, A, B, and C.  If we have 3 transient failures that cause us to reconnect (regardless of how those failures are distributed across the subchannels -- could be 3 failures on A, or one on A and two on B, or one failure on each), we will put the policy into shutdown, even if all of those reconnections are successful.

Conversely, if the new subchannel manages to reconnect immediately, the call to `grpc_subchannel_check_connectivity()` on line 666 could return `READY` immediately, in which case we would fail to increment the `num_ready` counter.

I'm not sure what the easiest way to fix this is.  My first thought was that we should simply call `update_state_counters_locked()` again right after we reset `curr_connectivity_state`.  But if we did that, then we'd never notice that all of the subchannels had gone into state `SHUTDOWN`, because they'd never stay there.  It might be that we need to explicitly track the set of subchannels that's currently trying to reconnect.

Also note that we should probably not be resetting `prev_connectivity_state` until after `update_state_counters_locked()` has a look at the current value -- in fact, we might even want to have `update_state_counters_locked()` update `prev_connectivity_state` after it's done updating the counters.